modules/nf-core/deseq2/differential/templates/deseq_de.r

```diff
@@ -0,0 +1,354 @@
+        paste0(
+        'Chosen contrast variable \"',
+        opt\$contrast_variable,
+        '\" not in sample sheet'
+        )
+    )
+} else if (any(!c(opt\$reflevel, opt\$treatlevel) %in% sample.sheet[[opt\$contrast_variable]])) {
+    stop(
+        paste(
+        'Please choose reference and treatment levels that are present in the',
+        opt\$contrast_variable,
+        'column of the sample sheet'
+        )
+    )
+} else if (!is.null(opt\$blocking)) {
+    blocking.vars = unlist(strsplit(opt\$blocking, split = ','))
+    if (!all(blocking.vars %in% colnames(sample.sheet))) {
+        stop(
+            paste0(
+                'One or more of the blocking variables specified (',
+                opt\$blocking,
+                ') do not correspond to sample sheet columns.'
+            )
+        )
+    }
+}
+
+# Now specify the model. Use cell-means style so we can be explicit with the
+# contrasts
+
+model <- '~ 0 +'
+
+if (!is.null(opt\$blocking_variables)) {
+    model <- paste(model, paste(blocking.vars, collapse = '+'))
+}
+
+# Make sure all the appropriate variables are factors
+
+for (v in c(blocking.vars, opt\$contrast_variable)) {
+    sample.sheet[[v]] <- as.factor(sample.sheet[[v]])
+}
+
+# Variable of interest goes last, see https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs
+
+model <- paste(model, opt\$contrast_variable, sep = ' + ')
+
+################################################
+################################################
+## Run DESeq2 processes                       ##
+################################################
+################################################
+
+dds <- DESeqDataSetFromMatrix(
+    countData = round(count.table),
+    colData = sample.sheet,
+    design = as.formula(model)
+)
+
+dds <- DESeq(
+    dds,
+    test = opt\$test,
+    fitType = opt\$fit_type,
+    minReplicatesForReplace = opt\$min_replicates_for_replace,
+    useT = opt\$use_t,
+    sfType = opt\$sf_type
+)
+
+comp.results <-
+    results(
+        dds,
+        lfcThreshold = opt\$lfc_threshold,
+        altHypothesis = opt\$alt_hypothesis,
+        independentFiltering = opt\$independent_filtering,
+        alpha = opt\$alpha,
+        pAdjustMethod = opt\$p_adjust_method,
+        minmu = opt\$minmu,
+        contrast = c(
+            opt\$contrast_variable,
+            c(opt\$treatment_level, opt\$reference_level)
+        )
+    )
+
+################################################
+################################################
+## Generate outputs                           ##
+################################################
+################################################
+
+# Common function to round numerics in data frames - mostly useful for testing,
+# since it doesn't seem to be possible to make outputs entirely reproducible
+# across machienes, even with set.seed(), but the differences are tiny.
+
+prepare_results <- function(x){
```

Comment:
I would just round to 8dp whatever, don't need more than that.

Reply:
That did cross my mind, but I came down on the side of keeping results 'as DESeq2 intended' except when we needed the rounding for testing. But happy to remove in the name of simplicity.